library/internal/buffer.go

```diff
@@ -0,0 +1,64 @@
+package internal
+
+import (
+	"bytes"
+	"io"
+)
+
+//MultiplexingBuffer is a type that will allow storing, replaying and streaming data from it's internal buffer
+//It implements io.Writer storing the written bytes to an unbound buffer internally
+//The Reader function will return an io.Reader that will stream data starting with the buffer, until the Close method of the underlying MultiplexingBuffer is called
+type MultiplexingBuffer struct {
+	closed bool
+	buffer []byte
+}
+
+type streamingReader struct {
+	multiplexingBuffer *MultiplexingBuffer
+	idx                int
+	drained            bool
+}
+
+func (s *streamingReader) Read(p []byte) (n int, err error) {
+
+	toRead := len(p)
+
+	if s.drained {
+		return 0, io.EOF
+	}
+
+	for i := 0; i < toRead; i++ {
+		if i+s.idx >= len(s.multiplexingBuffer.buffer) {
+			if s.multiplexingBuffer.closed {
+				s.drained = true
+			}
+			s.idx += i
+
+			return i, nil
+		}
```

Comment:
If whole content of buffer was read before and streaming reader is not closed each next Read call will return `0, nil`   Wonder how a caller knows when to retry the Read call ? 
`

Reply:
In my view this is exactly the scenario described in the io.Read docs:
![image](https://user-images.githubusercontent.com/5039095/127623969-7872e234-77a3-4ac8-aa81-276044f0d03c.png)

The app did not produce any output since the last read but it's still not finished hence len(p) == 0